homeassistant/components/buienradar/util.py

```diff
@@ -0,0 +1,228 @@
+"""Shared utilities for different supported platforms."""
+import asyncio
+from datetime import datetime, timedelta
+import logging
+
+import aiohttp
+import async_timeout
+
+from buienradar.buienradar import parse_data
+from buienradar.constants import (
+    ATTRIBUTION,
+    CONDITION,
+    CONTENT,
+    DATA,
+    FORECAST,
+    HUMIDITY,
+    MESSAGE,
+    PRESSURE,
+    STATIONNAME,
+    STATUS_CODE,
+    SUCCESS,
+    TEMPERATURE,
+    VISIBILITY,
+    WINDAZIMUTH,
+    WINDSPEED,
+)
+from buienradar.urls import JSON_FEED_URL, json_precipitation_forecast_url
+
+from homeassistant.const import CONF_LATITUDE, CONF_LONGITUDE
+from homeassistant.helpers.aiohttp_client import async_get_clientsession
+from homeassistant.helpers.event import async_track_point_in_utc_time
+from homeassistant.util import dt as dt_util
+
+
+from .const import SCHEDULE_OK, SCHEDULE_NOK
+
+
+_LOGGER = logging.getLogger(__name__)
+
+
+class BrData:
+    """Get the latest data and updates the states."""
+
+    def __init__(self, hass, coordinates, timeframe, devices):
+        """Initialize the data object."""
+        self.devices = devices
+        self.data = {}
+        self.hass = hass
+        self.coordinates = coordinates
+        self.timeframe = timeframe
+
+    async def update_devices(self):
+        """Update all devices/sensors."""
+        if self.devices:
+            tasks = []
+            # Update all devices
+            for dev in self.devices:
+                if dev.load_data(self.data):
+                    tasks.append(dev.async_update_ha_state())
+
+            if tasks:
+                await asyncio.wait(tasks)
+
+    async def schedule_update(self, minute=1):
+        """Schedule an update after minute minutes."""
+        _LOGGER.debug("Scheduling next update in %s minutes.", minute)
+        nxt = dt_util.utcnow() + timedelta(minutes=minute)
+        async_track_point_in_utc_time(self.hass, self.async_update, nxt)
+
+    async def get_data(self, url):
```

Comment:
For the future: This should be extracted to a standalone library published on pypi.